### ISSUE TITLE
Merge 10.3 (invisible columns)

### ISSUE THREAD

**Issue by midenok** [tempesta-tech/mariadb#407]
### Conflicted files
        both modified:   sql/field.cc
        both modified:   sql/share/errmsg-utf8.txt
        both modified:   sql/sql_base.cc
        both modified:   sql/sql_insert.cc
        both modified:   sql/sql_show.cc
        both modified:   sql/table.cc
        both modified:   sql/unireg.cc
### Important conflicts
```diff
--- a/sql/sql_base.cc
+++ b/sql/sql_base.cc
@@@ -8296,13 -8247,10 +8307,20 @@@ fill_record(THD *thd, TABLE *table, Fie
      /* Ensure that all fields are from the same table */
      DBUG_ASSERT(field->table == table);
  
++<<<<<<< HEAD
 +    if (table->versioned() && field->vers_sys_field() && !ignore_errors)
 +    {
 +      my_error(ER_NONUPDATEABLE_COLUMN, MYF(0), field->field_name.str);
 +      goto err;
 +    }
 +
 +    value=v++;
++=======
+     if (field->field_visibility != NOT_INVISIBLE)
+       continue;
+     else
+       value=v++;
++>>>>>>> 10.3
      if (field->field_index == autoinc_index)
        table->auto_increment_field_not_null= TRUE;
      if (field->vcol_info)
```
```diff
--- a/sql/sql_show.cc
+++ b/sql/sql_show.cc
@@@ -2183,17 -2162,18 +2183,26 @@@ int show_create_table(THD *thd, TABLE_L
    */
    old_map= tmp_use_all_columns(table, table->read_set);
  
+   bool not_the_first_field= false;
    for (ptr=table->field ; (field= *ptr); ptr++)
    {
+ 
      uint flags = field->flags;
  
++<<<<<<< HEAD
 +    if (vers_hide == VERS_HIDE_FULL &&
 +        (flags & (VERS_SYS_START_FLAG | VERS_SYS_END_FLAG)))
 +      continue;
 +
 +    if (ptr != table->field)
++=======
+     if (field->field_visibility > USER_DEFINED_INVISIBLE)
+        continue;
+     if (not_the_first_field)
++>>>>>>> 10.3
        packet->append(STRING_WITH_LEN(",\n"));
  
+     not_the_first_field= true;
      packet->append(STRING_WITH_LEN("  "));
      append_identifier(thd,packet,field->field_name.str,
                        field->field_name.length);
```
```diff
diff --cc sql/table.cc
index bf6084ccd86,f65d7fa123d..00000000000
--- a/sql/table.cc
+++ b/sql/table.cc
@@@ -1293,27 -1282,9 +1296,33 @@@ int TABLE_SHARE::init_from_binary_frm_i
          }
  #endif /*HAVE_SPATIAL*/
          break;
++<<<<<<< HEAD
 +      case EXTRA2_PERIOD_FOR_SYSTEM_TIME:
 +        if (system_period || length != 2 * sizeof(uint16))
 +          goto err;
 +        system_period = extra2;
 +        break;
 +      case EXTRA2_FIELD_FLAGS:
 +        if (extra2_field_flags)
 +          goto err;
 +        extra2_field_flags= extra2;
 +        extra2_field_flags_length= length;
 +        break;
 +      case EXTRA2_VTMD:
 +        if (vtmd_used)
 +          goto err;
 +        share->vtmd= *extra2;
 +        if (share->vtmd)
 +        {
 +          share->table_category= TABLE_CATEGORY_LOG;
 +          share->no_replicate= true;
 +        }
 +        vtmd_used= true;
++=======
+       case EXTRA2_FIELD_FLAGS:
+          field_properties = extra2;
+          field_additional_property_length= length;
++>>>>>>> 10.3
          break;
        default:
          /* abort frm parsing if it's an unknown but important extra2 value */
@@@ -1631,9 -1602,8 +1640,12 @@@
    memcpy(record, frm_image + record_offset, share->reclength);
  
    disk_buff= frm_image + pos + FRM_FORMINFO_SIZE;
- 
    share->fields= uint2korr(forminfo+258);
++<<<<<<< HEAD
 +  if (extra2_field_flags && extra2_field_flags_length != share->fields)
++=======
+   if (field_properties && field_additional_property_length != share->fields)
++>>>>>>> 10.3
      goto err;
    pos= uint2korr(forminfo+260);   /* Length of all screens */
    n_length= uint2korr(forminfo+268);
@@@ -2046,15 -1987,15 +2059,27 @@@
      reg_field->field_index= i;
      reg_field->comment=comment;
      reg_field->vcol_info= vcol_info;
++<<<<<<< HEAD
 +    reg_field->flags|= flags;
 +    if (extra2_field_flags)
 +    {
 +      uchar flags= *extra2_field_flags++;
 +      if (flags & VERS_OPTIMIZED_UPDATE)
 +        reg_field->flags|= VERS_UPDATE_UNVERSIONED_FLAG;
 +      if (flags & HIDDEN)
 +        reg_field->flags|= HIDDEN_FLAG;
 +    }
++=======
+     if(field_properties!=NULL)
+     {
+       uint temp= *field_properties++;
+       reg_field->field_visibility= f_visibility(temp);
+     }
+     if (reg_field->field_visibility == USER_DEFINED_INVISIBLE)
+       status_var_increment(thd->status_var.feature_invisible_columns);
+     if (reg_field->field_visibility == NOT_INVISIBLE)
+       share->visible_fields++;
++>>>>>>> 10.3
      if (field_type == MYSQL_TYPE_BIT && !f_bit_as_char(pack_flag))
      {
        null_bits_are_used= 1;
```

**Comment by midenok** [comment]
```diff
diff --cc sql/unireg.cc
index 5da14e31243,a0f42f68bf0..00000000000
--- a/sql/unireg.cc
+++ b/sql/unireg.cc
@@@ -87,44 -87,19 +87,60 @@@ static uchar *extra2_write(uchar *pos, 
    return extra2_write(pos, type, reinterpret_cast<LEX_CSTRING *>(str));
  }
  
++<<<<<<< HEAD
 +static const bool ROW_START = true;
 +static const bool ROW_END = false;
 +
 +inline
 +uint16
 +vers_get_field(HA_CREATE_INFO *create_info, List<Create_field> &create_fields, bool row_start)
 +{
 +  DBUG_ASSERT(create_info->versioned());
 +
 +  List_iterator<Create_field> it(create_fields);
 +  Create_field *sql_field = NULL;
 +
 +  const LString_i row_field= row_start ? create_info->vers_info.as_row.start
 +                                   : create_info->vers_info.as_row.end;
 +  DBUG_ASSERT(row_field);
 +
 +  for (unsigned field_no = 0; (sql_field = it++); ++field_no)
 +  {
 +    if (row_field == sql_field->field_name)
 +    {
 +      DBUG_ASSERT(field_no <= uint16(~0U));
 +      return uint16(field_no);
 +    }
 +  }
 +
 +  DBUG_ASSERT(0); /* Not Reachable */
 +  return 0;
 +}
 +
 +bool has_extra2_field_flags(List<Create_field> &create_fields)
 +{
 +  List_iterator<Create_field> it(create_fields);
 +  while (Create_field *f= it++)
 +  {
 +    if (f->flags & (VERS_UPDATE_UNVERSIONED_FLAG | HIDDEN_FLAG))
 +      return true;
 +  }
 +  return false;
++=======
+ static uchar *extra2_write_additional_field_properties(uchar *pos,
+                    int number_of_fields,List_iterator<Create_field> * it)
+ {
+   *pos++=EXTRA2_FIELD_FLAGS;
+   /*
+    always first 2  for field visibility
+   */
+   pos= extra2_write_len(pos, number_of_fields);
+   Create_field *cf;
+   while((cf=(*it)++))
+     *pos++= cf->field_visibility;
+   it->rewind();
+   return pos;
++>>>>>>> 10.3
  }
  
  /**
@@@ -258,23 -246,9 +287,29 @@@ LEX_CUSTRING build_frm_image(THD *thd, 
  
    if (gis_extra2_len)
      extra2_size+= 1 + (gis_extra2_len > 255 ? 3 : 1) + gis_extra2_len;
++<<<<<<< HEAD
 +
 +  if (create_info->versioned())
 +  {
 +    extra2_size+= 1 + 1 + 2 * sizeof(uint16);
 +  }
 +
 +  if (create_info->vtmd())
 +  {
 +    extra2_size+= 1 + 1 + 1;
 +  }
 +
 +  bool has_extra2_field_flags_= has_extra2_field_flags(create_fields);
 +  if (has_extra2_field_flags_)
 +  {
 +    extra2_size+=
 +        1 + (create_fields.elements <= 255 ? 1 : 3) + create_fields.elements;
 +  }
++=======
+   if(have_additional_field_properties)
+     extra2_size+=1 + (create_fields.elements > 255 ? 3 : 1) +
+         create_fields.elements;
++>>>>>>> 10.3
  
    key_buff_length= uint4korr(fileinfo+47);
  
@@@ -330,40 -304,8 +365,45 @@@
      pos+= gis_field_options_image(pos, create_fields);
    }
  #endif /*HAVE_SPATIAL*/
++<<<<<<< HEAD
 +
 +  if (create_info->versioned())
 +  {
 +    *pos++= EXTRA2_PERIOD_FOR_SYSTEM_TIME;
 +    *pos++= 2 * sizeof(uint16);
 +    int2store(pos, vers_get_field(create_info, create_fields, ROW_START));
 +    pos+= sizeof(uint16);
 +    int2store(pos, vers_get_field(create_info, create_fields, ROW_END));
 +    pos+= sizeof(uint16);
 +  }
 +
 +  if (create_info->vtmd())
 +  {
 +    *pos++= EXTRA2_VTMD;
 +    *pos++= 1;
 +    *pos++= 1;
 +  }
 +
 +  if (has_extra2_field_flags_)
 +  {
 +    *pos++= EXTRA2_FIELD_FLAGS;
 +    pos= extra2_write_len(pos, create_fields.elements);
 +    List_iterator<Create_field> it(create_fields);
 +    while (Create_field *field= it++)
 +    {
 +      uchar flags= 0;
 +      if (field->flags & VERS_UPDATE_UNVERSIONED_FLAG)
 +        flags|= VERS_OPTIMIZED_UPDATE;
 +      if (field->flags & HIDDEN_FLAG)
 +        flags|= HIDDEN;
 +      *pos++= flags;
 +    }
 +  }
 +
++=======
+   if (have_additional_field_properties)
+     pos=extra2_write_additional_field_properties(pos,create_fields.elements,&it);
++>>>>>>> 10.3
    int4store(pos, filepos); // end of the extra2 segment
    pos+= 4;
  
```

**Comment by midenok** [comment]
TODO:
VERS_HIDDEN_FLAG should be replaced with SYSTEM_INVISIBLE (or COMPLETELY_INVISIBLE?).
[#409]

**Comment by midenok** [comment]
Closed by 7fc6fc1c5009e89344f0b84b67093b52cc4ff312
